core/services/relay/evm/codec_entry_test.go

```diff
@@ -0,0 +1,170 @@
+package evm
+
+import (
+	"fmt"
+	"math/big"
+	"reflect"
+	"testing"
+
+	"github.com/ethereum/go-ethereum/accounts/abi"
+	"github.com/pkg/errors"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+
+	commontypes "github.com/smartcontractkit/chainlink-common/pkg/types"
+
+	"github.com/smartcontractkit/chainlink/v2/core/services/relay/evm/types"
+)
+
+func TestCodecEntry(t *testing.T) {
+	t.Run("basic types", func(t *testing.T) {
+		type1, err := abi.NewType("uint16", "", []abi.ArgumentMarshaling{})
+		require.NoError(t, err)
+		type2, err := abi.NewType("string", "", []abi.ArgumentMarshaling{})
+		require.NoError(t, err)
+		type3, err := abi.NewType("uint24", "", []abi.ArgumentMarshaling{})
+		require.NoError(t, err)
+		type4, err := abi.NewType("int24", "", []abi.ArgumentMarshaling{})
+		require.NoError(t, err)
+		entry := codecEntry{
+			Args: abi.Arguments{
+				{Name: "Field1", Type: type1},
+				{Name: "Field2", Type: type2},
+				{Name: "Field3", Type: type3},
+				{Name: "Field4", Type: type4},
+			},
+		}
+		require.NoError(t, entry.Init())
+		native := reflect.New(entry.nativeType)
+		iNative := reflect.Indirect(native)
+		iNative.FieldByName("Field1").Set(reflect.ValueOf(uint16(2)))
+		iNative.FieldByName("Field2").Set(reflect.ValueOf("any string"))
+		iNative.FieldByName("Field3").Set(reflect.ValueOf(big.NewInt( /*2^24 - 1*/ 16777215)))
+		iNative.FieldByName("Field4").Set(reflect.ValueOf(big.NewInt( /*2^23 - 1*/ 8388607)))
+		// native and checked point to the same item, even though they have different "types"
+		// they have the same memory layout so this is safe per unsafe casting rules, see unsafe.Pointer for details
+		checked := reflect.NewAt(entry.checkedType, native.UnsafePointer())
```

Comment:
Can you elaborate on this? If this is true, how is it being enforced? Why is this notion  of 'same' different from 'assignable' and 'convertible'?

Reply:
I'll add more in the comments, but they are structurally the same eg below.  This is enforced when they are created in codecEntry.  I can add comments to that effect in codecEntry.

```
type Native struct {
   X *big.Int 
   Y int32
}

type Checked struct {
  X *int128
  Y int32
}

```
type int128 big.Int

Reply:
CanConvert would return false, it's not built to check that

```
type Foo struct {
	T time.Time
}

type t time.Time

type Fooz struct {
	T t
}

func main() {
	v := reflect.ValueOf(&Foo{T: time.Unix(1, 2)})
	fmt.Println(v.CanConvert(reflect.TypeOf(Fooz{})))
}

```

Reply:
> This is enforced when they are created in codecEntry

What do you mean? Will the compiler throw an error if this symmetry is broken?

Reply:
Nope, it's "Unsafe" for a reason.  If you break it you can panic or get undefined behaviour.

It's equivalent of using `unsafe.Pointer`, but in reflection.

From the doc of unsafe.Pointer, I'm following the safe rule of their example because I know the underlying types are the same, and int the same order (since the two types are created in the same method):

```
// ...
// The following patterns involving Pointer are valid.
// ...
// (1) Conversion of a *T1 to Pointer to *T2.
//
// Provided that **T2 is no larger than T1 and that the two share an equivalent
// memory layout**, this conversion allows reinterpreting data of one type as
// data of another type. An example is the implementation of
// math.Float64bits:
//
//	func Float64bits(f float64) uint64 {
//		return *(*uint64)(unsafe.Pointer(&f))
//	}
```

I'm happy to move the `codecEntry` into the types package and to encapsulate the usage of unsafe to the strcut that does the actual creation.  Then it only happens in one place and I can document why it's safe there and where the structs are built up so people know not to change that behaviour.

Ie: I would make CheckedType exposed so you can create that, then have a method exposed to convert to Native type.  The checked type gives us extra type safety for ints that use big.Int because it'll assure that you don't go over the bounds (abi.Pack doesn't bound check) and I figure it's nice to have.

Reply:
>  because I know the underlying types are the same,

I'm not worried about documentation. I'm trying to understand what will fail when someone makes a change that breaks this equivalence. Do we have tests specifically for this? If not, then it could be an extremely confusing indirect failure that looks like arbitrary data corruption. It seems ripe for fuzz testing too.

Reply:
I do plan to add fuzzing.  I can also add some tests that check if the structures produced are always the same (can also fuzz that too).

Reply:
I also want to add more explict capture of panic in the codec.  Not related to this, but I learned that nil can cause Pack to panic, so I'll add a recover and fuzz test that too :).